src/iostream.c

```diff
@@ -548,6 +555,211 @@ StartChildProcess(const Char * dir, const Char * prg, Char * args[])
 }
 
 
+/****************************************************************************
+**
+*F  SyExecuteProcess( <dir>, <prg>, <in>, <out>, <args> ) . . . . new process
+**
+**  Start  <prg> in  directory <dir>  with  standard input connected to <in>,
+**  standard  output  connected to <out>   and arguments.  No  path search is
+**  performed, the return  value of the process  is returned if the operation
+**  system supports such a concept.
+*/
+#ifdef HAVE_POSIX_SPAWN
+
+UInt SyExecuteProcess(Char * dir, Char * prg, Int in, Int out, Char * args[])
+{
+    posix_spawn_file_actions_t file_actions;
+
+    // setup file actions
+    if (posix_spawn_file_actions_init(&file_actions)) {
+        PErr("SyExecuteProcess: posix_spawn_file_actions_init failed");
+        return -1;
+    }
+
+    /* if <in> is -1 open "/dev/null"                                  */
+    if (in == -1) {
+        if (posix_spawn_file_actions_addopen(&file_actions, 0, "/dev/null",
+                                             O_RDONLY, 0)) {
+            PErr("SyExecuteProcess: addopen(0, \"/dev/null\") failed");
+            posix_spawn_file_actions_destroy(&file_actions);
+            return -1;
+        }
+    }
+    else {
+        int tin = SyBufFileno(in);
+        if (tin != 0 &&
+            posix_spawn_file_actions_adddup2(&file_actions, tin, 0)) {
+            PErr("SyExecuteProcess: adddup2(child, 0) failed");
+            posix_spawn_file_actions_destroy(&file_actions);
+            return -1;
+        }
+    }
+
+    /* if <out> is -1 open "/dev/null"                                 */
+    if (out == -1) {
+        if (posix_spawn_file_actions_addopen(&file_actions, 1, "/dev/null",
+                                             O_WRONLY, 0)) {
+            PErr("SyExecuteProcess: addopen(0, \"/dev/null\") failed");
+            posix_spawn_file_actions_destroy(&file_actions);
+            return -1;
+        }
+    }
+    else {
+        int tout = SyBufFileno(out);
+        if (tout != 1 &&
+            posix_spawn_file_actions_adddup2(&file_actions, tout, 1)) {
+            PErr("SyExecuteProcess: adddup2(tout, 1) failed");
+            posix_spawn_file_actions_destroy(&file_actions);
+            return -1;
+        }
+    }
+
+    FreezeStdin = 1;
+
+    // spawn subprocess
+    pid_t pid;
+    if (posix_spawn_with_dir(&pid, prg, &file_actions, 0, args, environ,
+                             dir)) {
+        PErr("SyExecuteProcess: posix_spawn_with_dir failed");
+        return -1;
+    }
+
+    // wait for some action
```

Comment:
Note: the "classic" code also has some logic to "ignore a CTRL-C". And it disables our SIGCHLD handler temporarily. We may need to add one or both back here.